### PR TITLE
Expose extensibility points for opting out of tabs in the reference manager

### PR DESCRIPTION
**Pull request by RaulPerez1**
This allows consumers of the project system to opt out for #399 and #400

### CODE REVIEW

**Comment by RaulPerez1** [comment]
@srivatsn this is the same PR as the one you just approved, redid it as the other one had a merge conflict and it was showing 50+ commits after I did the merge.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+{
+    /// <summary>
+    /// Base reference context provider which abstract CPS overrides
+    /// </summary>
+    public class BaseReferenceContextProvider : IVsReferenceManagerUserAsync
+    {
+        /// <summary>
+        /// Value used to override the CPS provider
+        /// </summary>
+        public const int OverrideCPSProvider = 1;
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs]: Any reason for this to be public?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        /// </summary>
+        private Lazy<Lazy<IVsReferenceManagerUserAsync, IVsReferenceManagerUserComponentMetadataView>> nextHandler;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BaseReferenceContextProvider"/> class.
+        /// </summary>
+        [ImportingConstructor]
+        public BaseReferenceContextProvider(ConfiguredProject configuredProject)
+        {
+            this.VsReferenceManagerUsers = new OrderPrecedenceImportCollection<IVsReferenceManagerUserAsync, IVsReferenceManagerUserComponentMetadataView>(projectCapabilityCheckProvider: configuredProject);
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs]: We don't use `this.`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        export =>
+                            export.Metadata.OrderPrecedence < order.OrderPrecedence &&
+                            export.Metadata.ProviderContextIdentifier == user.ProviderContextIdentifier);
+            });
+        }
+
+        /// <summary>
+        /// Gets the next handler in the chain.
+        /// </summary>
+        private IVsReferenceManagerUserAsync NextHandler
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs]: Nit: you could make this one line with an expression-bodied getter.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            get
+            {
+                return nextHandler?.Value?.Value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the collection of reference provider contexts that can handle individual reference type operations.
+        /// </summary>
+        [ImportMany]
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs]: We don't do property injection: we do constructor parameter injection.

**Comment by RaulPerez1** [comment]
@333fred This on it's own does nothing interesting so unit tests are pointless here. We should have however some integration tests once we settle down on the capability implementation (on the SDK side). Those would be checking what shows up on the reference manager.

**Comment by 333fred** [comment]
Ok. Can we file a tracking bug for that so we don't forget to go back and write them?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs (hunk 4 above)
Comment (RaulPerez1) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/BaseReferenceContextProvider.cs]: Removed the import on the config (which was already being imported in the constructor anyway). This type (OrderPrecedenceImportCollection) needs to be done via a property or MEF complains. There's many uses of it in the codebase in the same manner.

**Comment by RaulPerez1** [comment]
Still need those integration tests for the integration bit of it (i.e. this overrides CPS providers). I'll do that as part of the SDK changes. After noticing that I had a bug on the capability definition for the WinRT provider I decided that it was worth adding unit tests for that part at this layer.

**Comment by RaulPerez1** [comment]
@333fred do you have any concerns about these changes?

**Pull request merged**